ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/OffsetBuilder.scala

```diff
@@ -4,72 +4,100 @@
 package com.daml.ledger.participant.state.kvutils
 
 import java.io.{ByteArrayInputStream, ByteArrayOutputStream, DataInputStream, DataOutputStream}
-
 import com.daml.ledger.offset.Offset
 
+import java.nio.ByteBuffer
+
 /** Helper functions for generating 16 byte [[Offset]]s from integers.
   * The created offset will look as follows:
-  * | highest index (64 bits) | middle index (32 bits) | lowest index (32 bits) |
+  * | version (8 bits) | highest index (56 bits) | middle index (32 bits) | lowest index (32 bits) |
   * Leading zeros will be retained when generating the resulting offset bytes.
   *
   * Example usage:
   *
-  *   - If you have one record per block then use [[OffsetBuilder.fromLong]] with a single argument, the block ID.
-  *   - If you may have multiple records per block then use [[OffsetBuilder.fromLong]] with the index within the block as the second argument.
+  *   - If you have one record per block then use [[OffsetBuilder.of]] with a single argument, the block ID.
+  *   - If you may have multiple records per block then use [[OffsetBuilder.of]] with the index within the block as the second argument.
   *
   * @see com.daml.ledger.offset.Offset
   * @see com.daml.ledger.participant.state.kvutils.api.KeyValueParticipantStateReader
   */
 object OffsetBuilder {
-  private[kvutils] val highestStart = 0
+  val DefaultVersion: Byte = 0
+
+  private[kvutils] val highestStart = 1
   private[kvutils] val middleStart = 8
   private[kvutils] val lowestStart = 12
   private[kvutils] val end = 16
 
   def onlyKeepHighestIndex(offset: Offset): Offset = {
+    val ver = version(offset)
     val highest = highestIndex(offset)
-    fromLong(highest)
+    of(ver, highest)
   }
 
   def dropLowestIndex(offset: Offset): Offset = {
-    val (highest, middle, _) = split(offset)
-    fromLong(highest, middle)
+    val (version, highest, middle, _) = split(offset)
+    of(version, highest, middle)
   }
 
   def setMiddleIndex(offset: Offset, middle: Int): Offset = {
-    val (highest, _, lowest) = split(offset)
-    fromLong(highest, middle, lowest)
+    val (version, highest, _, lowest) = split(offset)
+    of(version, highest, middle, lowest)
   }
 
   def setLowestIndex(offset: Offset, lowest: Int): Offset = {
-    val (highest, middle, _) = split(offset)
-    fromLong(highest, middle, lowest)
+    val (version, highest, middle, _) = split(offset)
+    of(version, highest, middle, lowest)
   }
 
-  def fromLong(first: Long, second: Int = 0, third: Int = 0): Offset = {
+  def of(version: Byte, first: Long, second: Int = 0, third: Int = 0): Offset = {
     val bytes = new ByteArrayOutputStream
     val stream = new DataOutputStream(bytes)
-    stream.writeLong(first)
+    stream.writeByte(version.toInt)
+    writeSevenByteLong(first, stream)
```

Comment:
Do we want to throw an exception if `first: Long` is too long?

Reply:
Yes, good idea. I think you just want to check if `first < 0`.

With tests, please.

Reply:
What do you think about using an unsigned long? Haven't ever used that but maybe it makes sense.

Reply:
Also, I think I would check the maximum value as well. WDYT?

Reply:
Scala/Java has unsigned longs?

Reply:
And yes, you're right, you need to check the maximum value too; I was getting bits and bytes confused.

Reply:
This didn't show up for me:

> I think returning an Option or Try would be better.

I'd prefer not to; it would complicate the API for an edge case that should, in theory, never ever happen.

Reply:
Exception at least, then, but not sure what are the consequences.

Reply:
> Scala/Java has unsigned longs?

Since Java 8, I think. Let me try it out and see how it looks.

Reply:
> Exception at least, then, but not sure what are the consequences.

Probably more pleasant than the previous consequences of exceeding the limit, which were to roll around to `Long.MinValue` (`-9223372036854775808`).